### PR TITLE
Start tracking Architectural Decisions (ADs)

### PR DESCRIPTION
**Pull request by pcattori**
# ↪️ Pull Request

We're doing a good job of tracking user-facing changes in the CHANGELOG and discussing architectural decisions in RFCs, but we need a place to record our architectural decisions i.e. the result of an RFC.

Let's keep a log of Architectural Decision Records (ADRs)!

@JuliaMalkin : Included you as a reviewer in case you wanted to weigh-in on writing style. These docs are contributor-only, so less user-facing than the rest of the docs. Not urgent to get your input, but wanted to include you 😄 

## 💻 Examples

For more, see [this blog post](https://understandlegacycode.com/blog/earn-maintainers-esteem-with-adrs/)

### CODE REVIEW

File: adr/0002-library-of-composable-functions.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+## Consequences
+
+Immutable types and pure functions make the code much easier to reason about,
+drastically cutting down the time to ramp up and debug.
+
+Note that not all types and functions *have* to be immutable and pure,
+but immutable types and pure functions should be the default.
+
+If there are good reasons to make exceptions, we can do so, but we should include comments to explain why this decision was taken.
```
Comment (pcattori) [inline adr/0002-library-of-composable-functions.md:65]:
```suggestion
If there are good reasons to make exceptions, we can do so, but we should include comments to explain why that exception was made.
```

File: docs/contributor-guide/style-guide.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Style Guide
 
+Architectural Decision Records (ADRs) track our important design desicions and architectural decisions.
```
Comment (skalish) [inline docs/contributor-guide/style-guide.md:3]:
```suggestion
Architectural Decision Records (ADRs) track our important design decisions and architectural decisions.
```

File: docs/contributor-guide/adr/0008-standardized-imports.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+baz.do_the_thing()
+
+# option 2: import value
+from foo.bar.bazaar import do_the_thing
+do_the_thing()
+```
+
+Not to mention that each of these styles may be done with relative imports (replacing `foo.bar` with `.bar` if the `bar` package is a sibling).
+
+Confusinly, Option 1.a and Option 1.b are _conceptually_ the same, but mechanically there are [subtle differences](https://stackoverflow.com/questions/24807434/imports-in-init-py-and-import-as-statement/24968941#24968941).
```
Comment (pcattori) [inline docs/contributor-guide/adr/0008-standardized-imports.md:31]:
```suggestion
Confusingly, Option 1.a and Option 1.b are _conceptually_ the same, but mechanically there are [subtle differences](https://stackoverflow.com/questions/24807434/imports-in-init-py-and-import-as-statement/24968941#24968941).
```

File: docs/contributor-guide/adr/0009-separate-types-and-functions.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+## Context
+
+Code must be organized to be compatible with:
+- Static type-checking via [mypy](https://github.com/python/mypy)
+- Runtime execution during normal usage and running tests via [pytest](https://docs.pytest.org/en/stable/)
+- Static doc generation via [sphinx-autodoc-typehints](https://github.com/agronholm/sphinx-autodoc-typehints)
+
+Additionally:
+- functions should be able to refer to any type
```
Comment (pcattori) [inline docs/contributor-guide/adr/0009-separate-types-and-functions.md:17]:
```suggestion
- Functions should be able to refer to any type
```

File: docs/contributor-guide/adr/0003-reproducibility.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+Manage dependencies via [poetry](https://python-poetry.org/).
+
+Define tests via [nox](https://nox.thea.codes/en/stable/).
+
+Run tests in automation/CI via [Github Actions](https://github.com/features/actions).
+
+## Consequences
+
+This solution let's us:
```
Comment (skalish) [inline docs/contributor-guide/adr/0003-reproducibility.md:27]:
```suggestion
This solution lets us:
```

File: docs/contributor-guide/adr/0008-standardized-imports.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+- Must `from foo import bar` instead of `import foo.bar as bar`
+- Must not import functions directly. Instead import the containing module and use `module.function(...)`
+- Must not use relative imports. Use absolute imports instead.
+
+## Consequences
+
+Standardized import style helps linter correctly order imports.
+
+Choosing import styles is a syntactic choice without semantic meaning.
+Removing this choice should speed up development and review.
```
Comment (ianbakst) [inline docs/contributor-guide/adr/0008-standardized-imports.md:47]: Fantastic!

**Pull request merged**